### PR TITLE
Feature/main physical formula

### LINKED ISSUE #7: Implement main physical formula
## Description
Implement functions that calculate the following formula and all needed partial formulas.
Image

## Acceptance Criteria
The Formula produces correct values for tau.

### PR DESCRIPTION
**Pull request by Madddiiiin**
### Summary

This PR adds the functions needed to calculate the lifetime $tau_{final}$ from the input data.

---

### PR checklist

- [x] I have written tests that fail without my changes. _Or I did not add them on purpose._
- [x] I have included instructions on how to test this, assuming they are not already apparent from the linked issue.

---

### Closes/Fixes/Related to

* Closes #7

### CODE REVIEW

File: napytau/core/polynomials.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        ndarray: Array of polynomial values evaluated at the given distance points.
+    """
+
+    # If no coefficients are provided, return a ndarray of zeros
+    if len(coefficients) == 0:
+        return zeros(array(len(distances)), dtype=float)
+
+    # Evaluate the polynomial sum at the given time points
+    result: ndarray = zeros_like(distances)
+    for i, c in enumerate(coefficients):
```
Comment (Benedikt-Brunner) [inline napytau/core/polynomials.py:30]: please try to avoid one letter variable names

File: napytau/core/polynomials.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        result += c * power(distances, i)
+    return result
```
Comment (Benedikt-Brunner) [inline napytau/core/polynomials.py:32]:
```suggestion

    return sum_at_measuring_distances
```

File: napytau/core/polynomials.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    Returns:
+        ndarray:
+        Array of the derivative values of the polynomial at the given distance points.
+    """
+    result: ndarray = zeros_like(distances)
+    for i, c in enumerate(coefficients):
+        if i > 0:
+            result += i * c * power(distances, (i - 1))
+    return result
```
Comment (Benedikt-Brunner) [inline napytau/core/polynomials.py:57]:
```suggestion

    return sum_of_derivative_at_measuring_distances
```

File: napytau/core/polynomials.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        coefficients (ndarray):
+        Array of polynomial coefficients [a_0, a_1, ..., a_n],
+        where the polynomial is P(t) = a_0 + a_1*t + a_2*t^2 + ... + a_n*t^n.
+
+    Returns:
+        ndarray:
+        Array of the derivative values of the polynomial at the given distance points.
+    """
+    result: ndarray = zeros_like(distances)
+    for i, c in enumerate(coefficients):
```
Comment (Benedikt-Brunner) [inline napytau/core/polynomials.py:54]: letters don't cost anything

File: napytau/core/tau.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            delta_unshifted_intensities,
+            initial_coefficients,
+            distances,
+            t_opt,
+            weight_factor,
+        )
+    )[0]
+
+    # calculate decay times using the optimized coefficients
+    tau_i: ndarray = (
```
Comment (Benedikt-Brunner) [inline napytau/core/tau.py:71]: just to highlight it in this file, i still don't think this is a great name

File: napytau/core/tau.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            doppler_shifted_intensities,
+            unshifted_intensities,
+            delta_doppler_shifted_intensities,
+            delta_unshifted_intensities,
+            initial_coefficients,
+            distances,
+            t_opt,
+            weight_factor,
+        )
+    )[0]
```
Comment (Benedikt-Brunner) [inline napytau/core/tau.py:68]: since the PR is massive i can't quite tell, is this used somewhere else?
If not the second part of the returned tuple seems useless

File: tests/core/delta_tau_test.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
+import unittest
```
Comment (Benedikt-Brunner) [inline tests/core/delta_tau_test.py:1]: if there is time, we could test some edge cases, but this is not essential

File: tests/core/delta_tau_test.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+    numpy_module_mock.array = np.array
+    numpy_module_mock.testing = np.testing
+    numpy_module_mock.ndarray = np.ndarray
+    return polynomial_module_mock, zeros_mock, numpy_module_mock
+
+
+class DeltaChiUnitTests(unittest.TestCase):
+    @staticmethod
+    def test_canCalculateJacobianMatrixOutOfTimesAndCoefficients():
+        """ " Can calculate Jacobian matrix out of distances and coefficients."""
```
Comment (Benedikt-Brunner) [inline tests/core/delta_tau_test.py:30]:
```suggestion
        """ Can calculate Jacobian matrix out of distances and coefficients."""
```

File: tests/core/polynomials_test.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+from napytau.core.polynomials import polynomial_sum_at_measuring_distances
+from napytau.core.polynomials import differentiated_polynomial_sum_at_measuring_distances
+from numpy import ndarray
+from numpy import array
+from numpy import testing
+
+
+class PolynomialsUnitTest(unittest.TestCase):
+    @staticmethod
+    def test_CalculationForValidPolynomial():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:11]: please add comments below the function definition like all other tests have

File: tests/core/polynomials_test.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
+import unittest
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:1]: Once again numpy isn't mocked

File: tests/core/tau_test.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
+import unittest
```
Comment (Benedikt-Brunner) [inline tests/core/tau_test.py:1]: numpy isn't mocked and the doc comments are missing, i will review the file further after this is fixed

**Review by Benedikt-Brunner — CHANGES_REQUESTED** [review]
As expected, hefty chonk PR => hefty chonk review

But overall good work 👍

File: tests/core/core_test.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply tests/core/core_test.py:1]: Don't know why I forgot that, but I fixed it.

File: napytau/core/core.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    Computes the final decay time (tau_final) and its associated uncertainty
+
+    Args:
+        tau_i (ndarray):
+        Array of individual decay times (tau_i) for each measurement
+        delta_tau_i (ndarray):
+        Array of uncertainties associated with each tau_i
+
+    Returns:
+        tuple: Weighted mean of tau (float) and its uncertainty (float)
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/core.py:19]: Before we remove all of these, I would wait for the meeting tomorrow, like we discussed.

File: napytau/core/core.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
+from numpy import ndarray
+from numpy import pow
+from numpy import sum
+from numpy import sqrt
+from typing import Tuple
+
+
+def calculate_tau_final(tau_i: ndarray, delta_tau_i: ndarray) -> Tuple[float, float]:
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/core.py:8]: I now named it tau_i_values and delta_tau_i_values, that way we should have both advantages.

File: napytau/core/core.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    weights: ndarray = 1 / pow(delta_tau_i, 2)
+
+    # Calculate the weighted mean of tau_i
+    weighted_mean: float = sum(weights * tau_i) / sum(weights)
+
+    # Calculate the uncertainty of the weighted mean
+    uncertainty: float = sqrt(1 / sum(weights))
+
+    # Edge case handling for empty input array
+    if tau_i.size == 0:
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/core.py:31]: Yes.

File: napytau/core/core.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        return -1, -1
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/core.py:32]: You are right and I changed that comment for an explanation. I chose -1 instead of 0, because 0 could be mistaken for an actual, plausible result, while -1 clearly shows that the input doesn't make much sense.

File: pyproject.toml [hunk 17]
```diff
@@ -7,6 +7,8 @@ requires-python = ">=3.12"
 dependencies = [
     "customtkinter>=5.2.2",
     "matplotlib>=3.10.0",
+    "numpy>=2.1.3",
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply pyproject.toml:10]: This wasn't done deliberately, but rather automatically by the IDE. I changed it to 2.2.0 now.

File: napytau/core/chi.py [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+from numpy import sum
+from numpy import ndarray
+from numpy import mean
+from numpy import power
+from scipy.optimize import minimize
+from scipy.optimize import OptimizeResult
+from typing import Tuple
+
+
+# Chi^2 Funktion für festes t-hyp
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/chi.py:12]: I think this was one of the very first comments of this entire PR and probably only meant for myself lol. Fixed it.

File: napytau/core/chi.py [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+        delta_unshifted_intensities (ndarray):
+        Uncertainties in unshifted intensities
+        coefficients (ndarray):
+        Polynomial coefficients for fitting
+        distances (ndarray):
+        Array of distance points
+        t_hyp (float):
+        Hypothesis value for the scaling factor
+        weight_factor (float):
+        Weighting factor for unshifted intensities
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/chi.py:42]: Same as above with the docstrings in general. It will also automatically fix itself, when we implement the use of your custom datatypes.

File: napytau/core/chi.py [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+            unshifted_intensities,
+            delta_doppler_shifted_intensities,
+            delta_unshifted_intensities,
+            coefficients,
+            distances,
+            t_hyp,
+            weight_factor,
+        ),
+        initial_coefficients,
+        method="L-BFGS-B",  # Optimization method for bounded optimization
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/chi.py:119]: I now explained what makes this algorithm special, but honestly, I didn't do too much research on what algorithm would be best here.

File: napytau/core/chi.py [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+        lambda coefficients: chi_squared_fixed_t(
+            doppler_shifted_intensities,
+            unshifted_intensities,
+            delta_doppler_shifted_intensities,
+            delta_unshifted_intensities,
+            coefficients,
+            distances,
+            t_hyp,
+            weight_factor,
+        ),
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/chi.py:117]: I'm not sure how to do this, since I think we need a lambda function here. Maybe you can give a suggestion.

[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/chi.py:117]: The only thing I can think of is changing it back to a regular function (with a descriptive name), like it once was.

File: napytau/core/chi.py [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+        initial_coefficients,
+        t_hyp,
+        weight_factor,
+    )[1]
+
+    # minimize chi-squared function over the range of t_hyp
+    result: OptimizeResult = minimize(
+        chi_squared_t_hyp,
+        x0=mean(t_hyp_range),  # Initial guess for t_hyp
+        bounds=[(t_hyp_range[0], t_hyp_range[1])],  # Boundaries for optimization
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/chi.py:178]: I now explained x0, but I think the bounds are self-explanatory.

File: napytau/core/chi.py [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+    )
+
+    # Return optimized t_hyp value
+    optimized_t_hyp: float = result.x
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/chi.py:182]: I only added this to make the lint tool happy. Otherwise it would complain that result.x has no static type.

File: napytau/core/core.py [hunk 24]
```diff
@@ -0,0 +1,35 @@
+from numpy import ndarray
```
[the remark this answers is not in the document]
Comment (Benedikt-Brunner) [reply napytau/core/core.py:1]: Please don't call it `core_main`, that says nothing. Have a look at the main file of the import/export module, the functions should still be descriptive.

File: napytau/core/chi.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (Benedikt-Brunner) [reply napytau/core/chi.py:117]: Didn't we turn off a lint rule just to allow this?

```python
whatever_function_name = lambda coefficients: chi_squared_fixed_t(
            doppler_shifted_intensities,
            unshifted_intensities,
            delta_doppler_shifted_intensities,
            delta_unshifted_intensities,
            coefficients,
            distances,
            t_hyp,
            weight_factor,
        )
```

File: napytau/core/core.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/core.py:1]: Then I'll call it calculate_lifetime i guess, cause that's what it does. I'm kind of running out of names.

File: tests/core/chi_test.py [hunk 25]
```diff
@@ -0,0 +1,640 @@
...
+
+def set_up_mocks() -> (MagicMock, MagicMock):
+    polynomials_mock = MagicMock()
+    polynomials_mock.polynomial_sum_at_measuring_times = MagicMock()
+    polynomials_mock.differentiated_polynomial_sum_at_measuring_times = MagicMock()
+
+    scipy_optimize_module_mock = MagicMock()
+    scipy_optimize_module_mock.minimize = MagicMock()
+
+    return polynomials_mock, scipy_optimize_module_mock
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply tests/core/chi_test.py]: This was a bit dumb and also took quite some time to fix. Hopefully it is okay now. Unfortunately this brought the chi_test file to almost 1000 lines of code, but I don't really know, how we could have avoided that. (800 lines are for one function alone)

File: tests/core/chi_test.py [hunk 26]
```diff
@@ -0,0 +1,640 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply tests/core/chi_test.py:1]: Like I said above, from the now 930 lines, 800 are for one function. I don't really know, how we could split that up.

[the remark this answers is not in the document]
Comment (Madddiiiin) [reply tests/core/chi_test.py:1]: Also, SciPy should be mocked for the functions that actually use it.

File: napytau/core/delta_tau.py [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+    jacobian_matrix: ndarray = zeros((len(distances), len(coefficients)))
+
+    epsilon: float = 1e-8  # small disturbance value
+
+    # Loop over each coefficient and calculate the partial derivative
+    for i in range(len(coefficients)):
+        perturbed_coefficients: ndarray = array(coefficients, dtype=float)
+        perturbed_coefficients[i] += epsilon  # slightly disturb the current coefficient
+
+        # Compute the disturbed and original polynomial values at the given times
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/delta_tau.py:36]: Forgot to change that, when I renamed times to distances. Fixed it.

[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/delta_tau.py:36]: I also found quite a few more of those in this file.

File: napytau/core/delta_tau.py [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+        perturbed_function: ndarray = polynomial_sum_at_measuring_distances(
+            distances, perturbed_coefficients
+        )
+        original_function: ndarray = polynomial_sum_at_measuring_distances(
+            distances, coefficients
+        )
+
+        # Calculate the partial derivative coefficients and store it in the
+        # Jacobian matrix
+        jacobian_matrix[:, i] = (perturbed_function - original_function) / epsilon
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/delta_tau.py:46]: For now I added a comment, explaining the syntax. Do you think that is enough or should we use an entirely different syntax here?

File: napytau/core/delta_tau.py [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+    covariance_matrix: ndarray = calculate_covariance_matrix(
+        delta_shifted_intensities, distances, coefficients
+    )
+
+    # Calculate the polynomial uncertainty contributions
+    for k in range(len(coefficients)):
+        for l in range(len(coefficients)):  # noqa E741
+            delta_p_j_i_squared = (
+                delta_p_j_i_squared
+                + power(distances, k) * power(distances, l) * covariance_matrix[k, l]
```
[the remark this answers is not in the document]
Comment (Madddiiiin) [reply napytau/core/delta_tau.py:128]: We should probably check that the dimensions always match. It should be true, because the shape of the jacobian_matrix is (len(distances), len(coefficients)), the shape of the weight_matrix is (len(coefficients), len(distances) and so the shape of the fit_matrix should always be (len(coefficients), len(coefficients)).

File: tests/core/delta_tau_test.py [hunk 30]
```diff
@@ -0,0 +1,387 @@
...
+            )
+
+            np.testing.assert_allclose(
+                numpy_module_mock.linalg.inv.mock_calls[0].args[0],
+                np.array(
+                    [[3.81250000e16, 1.27083333e16], [1.27083333e16, 4.23611111e15]]
+                ),
+            )
+
+    def testCanCalculateErrorPropagation(self):
```
Comment (Benedikt-Brunner) [inline tests/core/delta_tau_test.py:167]:
```suggestion
    def testCanCalculateTheErrorPropagation(self):
```

File: tests/core/polynomials_test.py [hunk 31]
```diff
@@ -0,0 +1,329 @@
...
+            # At x = 2: 2 + 3(2) + 4(2^2) = 2 + 6 + 16 = 24
+            # At x = 3: 2 + 3(3) + 4(3^2) = 2 + 9 + 36 = 47
+            expected_result: ndarray = array([9, 24, 47])
+            testing.assert_array_equal(
+                evaluate_polynomial_at_measuring_distances(distances, coefficients),
+                expected_result,
+            )
+
+    @staticmethod
+    def test_CanEvaluateAPolynomialAtMeasuringDistancesForEmptyDistanceInput():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:54]: The test name does not match the doc comment, make sure it does everywhere

File: tests/core/polynomials_test.py [hunk 32]
```diff
@@ -0,0 +1,329 @@
...
+            coefficients: ndarray = array([2, 3, 4])
+            # With an empty input array, the result should also be an empty array
+            expected_result: ndarray = array([])
+            testing.assert_array_equal(
+                evaluate_polynomial_at_measuring_distances(distances, coefficients),
+                expected_result,
+            )
+
+    @staticmethod
+    def test_CanEvaluateAPolynomialAtMeasuringDistancesForSingleDistanceMeasurement():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:82]:
```suggestion
    def test_CanEvaluateAPolynomialAtMeasuringDistancesForASingleDistanceMeasurement():
```
```suggestion
    def test_CanEvaluateAPolynomialAtMeasuringDistancesForASingleDistance():
```

File: tests/core/polynomials_test.py [hunk 33]
```diff
@@ -0,0 +1,329 @@
...
+            # Polynomial: f(x) = 1 + 2x
+            # At x = 2: 1 + 2(2) = 5
+            expected_result: ndarray = array([5])
+            testing.assert_array_equal(
+                evaluate_polynomial_at_measuring_distances(distances, coefficients),
+                expected_result,
+            )
+
+    @staticmethod
+    def test_CanEvaluateASingleValuePolynomialAtMeasuringDistances():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:111]: This is probably trying to say, a polynomial of degree one. The function always runs on a single polynomial

File: tests/core/polynomials_test.py [hunk 34]
```diff
@@ -0,0 +1,329 @@
...
+            # Constant polynomial: f(x) = 5
+            # All values should be 5
+            expected_result: ndarray = array([5, 5, 5])
+            testing.assert_array_equal(
+                evaluate_polynomial_at_measuring_distances(distances, coefficients),
+                expected_result,
+            )
+
+    @staticmethod
+    def test_CanEvaluateAnEmptyPolynomialAtMeasuringDistances():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:144]: Again, i think describing the polynomial by it's degree is more intuitive

File: tests/core/polynomials_test.py [hunk 35]
```diff
@@ -0,0 +1,329 @@
...
+            },
+        ):
+            from napytau.core.polynomials import (
+                evaluate_polynomial_at_measuring_distances,
+            )
+
+            distances: ndarray = array([1, 2])
+            coefficients: ndarray = array([])
+            # With an empty coefficients array, the result should be an array of zeros
+            # with the same length as 'distances'
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:164]: One could argue, that an empty polynomial should lead to an error, as it is not a valid function. A polynomial with only one coefficient that is 0 would be a valid function:

```
f(x) = 
vs
f(x) = 0
```

Comment (Benedikt-Brunner) [reply tests/core/polynomials_test.py:164]: => i would expect the function to throw an error in this case, consider adding a `CoreError` for this.

File: tests/core/polynomials_test.py [hunk 36]
```diff
@@ -0,0 +1,329 @@
...
+            expected_result: ndarray = array([11, 19, 27])
+            testing.assert_array_equal(
+                evaluate_differentiated_polynomial_at_measuring_distances(
+                    distances, coefficients
+                ),
+                expected_result,
+            )
+
+    @staticmethod
+    def test_CanEvaluateADifferentiatedPolynomialAtMeasuringDistancesForEmptyDistanceInput():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:206]: Also doesn't match the doc comment

File: tests/core/polynomials_test.py [hunk 37]
```diff
@@ -0,0 +1,329 @@
...
+            expected_result: ndarray = array([])
+            testing.assert_array_equal(
+                evaluate_differentiated_polynomial_at_measuring_distances(
+                    distances, coefficients
+                ),
+                expected_result,
+            )
+
+    @staticmethod
+    def test_CanEvaluateADifferentiatedPolynomialAtMeasuringDistancesForSingleDistanceMeasurement():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:236]: see above

File: tests/core/polynomials_test.py [hunk 38]
```diff
@@ -0,0 +1,329 @@
...
+            expected_result: ndarray = array([0, 0, 0])
+            testing.assert_array_equal(
+                evaluate_differentiated_polynomial_at_measuring_distances(
+                    distances, coefficients
+                ),
+                expected_result,
+            )
+
+    @staticmethod
+    def test_CanEvaluateADifferentiatedEmptyPolynomialAtMeasuringDistances():
```
Comment (Benedikt-Brunner) [inline tests/core/polynomials_test.py:298]: See my comments about empty polynomials above

File: tests/core/tau_final_test.py [hunk 39]
```diff
@@ -0,0 +1,67 @@
...
+            tau_i: ndarray = array([2, 4])
+            delta_tau_i: ndarray = array([1, 2])
+            expected_tau_final: float = 2.4
+            expected_uncertainty: float = 0.894427191
+            tau_final: (float, float) = calculate_tau_final(tau_i, delta_tau_i)
+            self.assertAlmostEqual(tau_final[0], expected_tau_final)
+            self.assertAlmostEqual(tau_final[1], expected_uncertainty)
+
+    def test_calculateTauFinalForEmptyInput(self):
+        """Returns (-1, -1) if input arrays are empty"""
```
Comment (Benedikt-Brunner) [inline tests/core/tau_final_test.py:42]: The name and the comment have nothing to do with each other 💀

File: tests/core/tau_final_test.py [hunk 40]
```diff
@@ -0,0 +1,67 @@
...
+    numpy_module_mock = MagicMock()
+    numpy_module_mock.power = MagicMock()
+    numpy_module_mock.sum = MagicMock()
+    numpy_module_mock.sqrt = MagicMock()
+    return numpy_module_mock
+
+
+class TauFinalUnitTest(unittest.TestCase):
+    def test_calculateTauFinalForValidData(self):
+        """Calculates the weighted mean and the uncertainty of tau correctly"""
```
Comment (Benedikt-Brunner) [inline tests/core/tau_final_test.py:17]: 🤯

File: tests/core/tau_final_test.py [hunk 41]
```diff
@@ -0,0 +1,67 @@
...
+        ):
+            from napytau.core.tau_final import calculate_tau_final
+
+            tau_i: ndarray = array([2, 4])
+            delta_tau_i: ndarray = array([1, 2])
+            expected_tau_final: float = 2.4
+            expected_uncertainty: float = 0.894427191
+            tau_final: (float, float) = calculate_tau_final(tau_i, delta_tau_i)
+            self.assertAlmostEqual(tau_final[0], expected_tau_final)
+            self.assertAlmostEqual(tau_final[1], expected_uncertainty)
```
Comment (Benedikt-Brunner) [inline tests/core/tau_final_test.py:39]:
```suggestion
            tau_i: ndarray = array([2, 4])
            delta_tau_i: ndarray = array([1, 2])
            
             tau_final: (float, float) = calculate_tau_final(tau_i, delta_tau_i)
            
            expected_tau_final: float = 2.4
            expected_uncertainty: float = 0.894427191
            self.assertAlmostEqual(tau_final[0], expected_tau_final)
            self.assertAlmostEqual(tau_final[1], expected_uncertainty)
```

Always try to aim for a structure of:

1. Setup
2. Execution
3. Assertion

When writing tests

File: tests/core/tau_final_test.py [hunk 42]
```diff
@@ -0,0 +1,67 @@
...
+            tau_i: ndarray = array([])
+            delta_tau_i: ndarray = array([])
+            expected_tau_final: float = -1
+            expected_uncertainty: float = -1
+            self.assertEqual(
+                calculate_tau_final(tau_i, delta_tau_i)[0], expected_tau_final
+            )
+            self.assertEqual(
+                calculate_tau_final(tau_i, delta_tau_i)[1], expected_uncertainty
+            )
```
Comment (Benedikt-Brunner) [inline tests/core/tau_final_test.py:67]: See above, also why is the function called twice?

**Review by Benedikt-Brunner — CHANGES_REQUESTED** [review]
Nice work 👍 

The actual code looks really good already, there are some inconsistencies and touchups left in the tests.

File: tests/core/polynomials_test.py (hunk 35 above)
Comment (Madddiiiin) [reply tests/core/polynomials_test.py:164]: I agree, I think I mixed up 0 parameters and a function of degree 0 in my head.

**Review by Benedikt-Brunner — APPROVED** [review]
LGTM 🎉 

I did some final touch up, so we can avoid another round of review
